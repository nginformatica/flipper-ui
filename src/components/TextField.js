import React from 'react'
import _ from 'prop-types'
import styled from 'styled-components'
import { background, primary, text } from '../colors'

const StyledTextField = styled.input`
    outline: none;
    padding: 8px;
    border-radius: 6px;
    line-height: 1.5em;
    margin: 1em;
    border: 1px solid ${background.dark};
    color: ${text};
    font-size: 14px;
    font-family: 'Roboto', sans-serif;
    transition: all 500ms ease;
    height: 24px;
    &:focus {
        border: 1px solid ${primary.normal};
        border-radius: 6px;
    }
`

const TextField = props => <StyledTextField { ...props } />

TextField.propTypes = {
    placeholder: _.string,
    value: _.string,
    style: _.object,
    onChange: _.func
}

export default TextField
