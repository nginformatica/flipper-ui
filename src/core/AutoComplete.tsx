import Downshift from 'downshift'
import {
    contains,
    filter,
    is,
    isNil,
    pipe,
    propEq,
    propOr,
    toLower,
    uniq,
    unless,
    when
} from 'ramda'
import React, {
    cloneElement,
    Component,
    FocusEvent,
    Fragment,
    ReactElement,
    ReactNode
} from 'react'
import Paper from './Paper'

interface IProps {
    data: object[] | string[]
    value: string
    defaultValue?: string
    inputElement: ReactElement<any>
    InputProps?: object
    defaultIsOpen?: boolean
    openOnFocus?: boolean
    renderSuggestion: (
        suggestion: string | object,
        itemProps: object,
        selected: boolean
    ) => ReactNode
    onChange?: (value: string) => void
    onSelect?: (value: string) => void
    onFocus?: (event: FocusEvent<HTMLInputElement>) => void
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void
}

class AutoComplete extends Component<IProps> {
    public autocomplete = { offsetWidth: 256 }
    public getSuggestions(inputValue) {
        if (this.props.openOnFocus && !inputValue) {
            return this.props.data
        }

        const items = inputValue
            ? this.props.data
            : []

        return uniq(
            filter(
                unless(
                    propEq('subheader', true),
                    pipe(
                        when(is(Object), propOr('', 'label')),
                        toLower,
                        contains(toLower(inputValue || ''))
                    )
                ),
                items
            )
        )
    }

    public handleSelect(value) {
        if (this.props.onSelect) {
            this.props.onSelect(value)
        } else {
            this.handleChange(value)
        }
    }

    public handleChange(value) {
        if (!isNil(value) && this.props.onChange) {
            this.props.onChange(value)
        }
    }

    public renderInput(props: {
        inputProps: object,
        onFocus: (event: FocusEvent<HTMLInputElement>) => void
    }) {
        return cloneElement(this.props.inputElement, {
            InputProps: {
                ...props,
                ...this.props.InputProps
            },
            inputProps: {
                ref: self => {
                    this.autocomplete = self
                }
            }
        })
    }

    public render() {
        const paperStyle = {
            position: 'absolute',
            width: this.autocomplete.offsetWidth,
            zIndex: 1099
        }

        return (
            <Downshift
                inputValue={ this.props.value }
                defaultInputValue={ this.props.defaultValue }
                defaultIsOpen={ this.props.defaultIsOpen }
                itemToString={ item => is(Object, item) ? item.label : item }
                onSelect={ this.handleSelect.bind(this) }
                onInputValueChange={ this.props.onChange }>
                {
                    ({ isOpen, getInputProps, inputValue, getItemProps, highlightedIndex, openMenu }) =>
                        <div>
                            {
                                this.renderInput(getInputProps({
                                    onBlur: this.props.onBlur,
                                    onFocus: (event: FocusEvent<HTMLInputElement>) => this.props.onFocus
                                        ? this.props.onFocus(event)
                                        : this.props.openOnFocus
                                            ? openMenu()
                                            : null,
                                    value: inputValue || ''
                                }))
                            }
                            {
                                isOpen && (
                                    <Paper square style={ paperStyle }>
                                        {
                                            this.getSuggestions(inputValue).map((suggestion, index) =>
                                                <Fragment key={ index }>
                                                    {
                                                        this.props.renderSuggestion(
                                                            suggestion,
                                                            getItemProps({ item: suggestion }),
                                                            highlightedIndex === index
                                                        )
                                                    }
                                                </Fragment>
                                            )
                                        }
                                    </Paper>
                                )
                            }
                        </div>
                }
            </Downshift>
        )
    }
}

export default AutoComplete
