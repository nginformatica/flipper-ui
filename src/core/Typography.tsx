import React from 'react'
import { DefaultProps } from './types'
import MuiTypography, { TypographyProps } from '@material-ui/core/Typography'

const Typography = ({
    children,
    margin,
    padding,
    style = {},
    variant = 'body2',
    ...otherProps
}: TypographyProps & DefaultProps) => (
    <MuiTypography
        {...otherProps}
        variant={variant}
        style={{ margin, padding, ...style }}>
        {children}
    </MuiTypography>
)

export default Typography
