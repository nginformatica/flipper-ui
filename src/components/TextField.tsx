import {
    createStyles,
    WithStyles,
    withStyles
} from '@material-ui/core'
import { TextField as MuiTextField } from '@material-ui/core'
import { merge } from 'ramda'
import React from 'react'
import { background, primary } from '../colors'

interface IProps extends WithStyles<typeof styles> {
    autoComplete?: string
    autoFocus?: boolean
    defaultValue?: string | number
    disabled?: boolean
    error?: boolean
    fullWidth?: boolean
    id?: string
    label?: string
    placeholder?: string
    style?: object
    margin?: 'none' | 'dense' | 'normal'
    multiline?: boolean
    name?: string
    required?: boolean
    select?: boolean
    type?: string
    value?: string | number | boolean | string[]
    InputProps?: object
    InputLabelProps?: object
    SelectProps?: object
    onChange?: (event) => {}
}

const styles = theme => createStyles({
    input: {
        '&:focus': {
            borderColor: theme.palette.primary.light,
            boxShadow: `0 0 0 0.2rem ${theme.palette.primary.light}D9`,
        },
        'backgroundColor': theme.palette.common.white,
        'border': `1px solid ${background.dark}`,
        'borderRadius': 4,
        'fontSize': 16,
        'padding': '10px 12px',
        'transition': theme.transitions.create(['border-color', 'box-shadow']),
        'width': 'calc(100% - 24px)'
    },
    label: {
        fontSize: 18
    },
    root: {
        'label + &': {
            marginTop: theme.spacing.unit * 3,
        },
        'padding': 0
    }
})

const TextField = ({ classes, InputProps, InputLabelProps, ...otherProps }: IProps) =>
    <MuiTextField
        { ...otherProps }
        InputProps={
            merge(
                {
                    classes: {
                        input: classes.input,
                        root: classes.root
                    },
                    disableUnderline: true
                },
                InputProps
            )
        }
        InputLabelProps={
            merge({
                    className: classes.label,
                    shrink: true
                },
                InputLabelProps
            )
        }
    />

export default withStyles(styles)(TextField)
