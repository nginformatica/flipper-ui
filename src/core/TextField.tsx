import {
    InputAdornment,
    TextField as MuiTextField,
    IconButton as MuiButton
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React, {
    ChangeEvent,
    KeyboardEvent,
    FC,
    FocusEvent,
    ReactNode,
    MouseEvent
} from 'react'
import { DefaultProps } from './types'
import { Clear, Help as ContactSupportIcon } from '@material-ui/icons'
import IconButton from './IconButton'
import styled from 'styled-components'

export interface TextFieldProps extends DefaultProps {
    autoComplete?: string
    autoFocus?: boolean
    defaultValue?: string | number
    disabled?: boolean
    error?: boolean
    fullWidth?: boolean
    id?: string
    label?: string
    placeholder?: string
    multiline?: boolean
    name?: string
    required?: boolean
    select?: boolean
    type?: string
    value?: string | number
    variant?: 'standard' | 'outlined' | 'filled'
    inputRef?: React.Ref<HTMLInputElement>
    inputProps?: object
    InputProps?: object
    InputLabelProps?: object
    SelectProps?: object
    rows?: string | number
    rowsMax?: string | number
    helperText?: ReactNode
    helperIcon?: ReactNode
    hasClear?: boolean
    onClear?: () => void
    onHelperClick?: () => void
    onChange?: (event: ChangeEvent<HTMLInputElement>) => void
    onPaste?: (event: React.ClipboardEvent<HTMLInputElement>) => void
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void
    onKeyUp?: (event: KeyboardEvent) => void
    onKeyDown?: (event: KeyboardEvent) => void
}

interface IHelperProps extends Pick<TextFieldProps, 'helperIcon'> {
    onHelperClick: (event: MouseEvent<HTMLButtonElement>) => void
}

const Helper = styled.div`
    width: 42px;
    height: 38px;
`

export const TextFieldWrapper = styled.div`
    display: flex;
    flex-direction: rows;
    width: 100%;
    align-items: center;
    button {
        display: none;
    }
    :hover button {
        display: flex;
    }
`

export const StaticTextFieldWrapper = styled.div`
    display: flex;
    flex-direction: rows;
    width: 100%;
    align-items: center;
`

export const useStyles = makeStyles({
    input: {
        fontSize: '14px',
        padding: '10px',
        height: 'auto'
    },
    outlinedInput: {
        fontSize: '14px',
        padding: '10px',
        height: 'auto'
    },
    outlinedLabel: {
        fontSize: '14px',
        transform: 'translate(14px, 13px) scale(1)'
    },
    outlinedMultiline: {
        padding: '0px'
    },
    iconOutlined: {
        right: '2px'
    }
})

export const HelperBox = (props: IHelperProps) => (
    <Helper>
        <IconButton padding='6px 2px' onClick={props.onHelperClick}>
            {props.helperIcon || <ContactSupportIcon color='primary' />}
        </IconButton>
    </Helper>
)

const renderEndAdornment = (onClear?: () => void) => (
    <InputAdornment position='end'>
        <MuiButton onClick={onClear} size='small'>
            <Clear style={{ fontSize: '15px' }} />
        </MuiButton>
    </InputAdornment>
)

const TextField: FC<TextFieldProps> = ({
    margin,
    padding,
    style = {},
    error,
    variant = 'outlined',
    InputLabelProps = {},
    InputProps = {},
    SelectProps = {},
    autoComplete = 'off',
    onHelperClick,
    helperIcon,
    fullWidth,
    hasClear,
    onClear,
    ...otherProps
}) => {
    const clearStyle = makeStyles({
        iconOutlined: {
            position: 'relative',
            right: '2px',
            marginLeft: '-20px'
        }
    })

    const hasValue = !!otherProps.value

    const endAdornment =
        hasValue && hasClear
            ? { endAdornment: renderEndAdornment(onClear) }
            : {}

    const classes = useStyles()
    const clearClass = clearStyle()

    const handleClick = () => {
        if (onHelperClick) {
            onHelperClick()
        }
    }

    const Wrapper = hasClear ? StaticTextFieldWrapper : TextFieldWrapper

    return (
        <Wrapper>
            <MuiTextField
                fullWidth={fullWidth}
                autoComplete={autoComplete}
                error={error}
                variant={variant as 'outlined'}
                style={{
                    margin,
                    padding,
                    ...style
                }}
                InputLabelProps={{
                    classes: {
                        outlined:
                            variant === 'outlined' ? classes.outlinedLabel : ''
                    },
                    ...InputLabelProps
                }}
                InputProps={{
                    classes: {
                        input:
                            variant === 'outlined' ? classes.outlinedInput : '',
                        multiline:
                            variant === 'outlined'
                                ? classes.outlinedMultiline
                                : ''
                    },
                    ...InputProps
                }}
                SelectProps={{
                    classes: {
                        iconOutlined: hasClear
                            ? clearClass.iconOutlined
                            : classes.iconOutlined
                    },
                    ...endAdornment,
                    ...SelectProps
                }}
                {...otherProps}
            />
            {onHelperClick && (
                <HelperBox
                    helperIcon={helperIcon}
                    onHelperClick={handleClick}
                />
            )}
        </Wrapper>
    )
}

export default TextField
