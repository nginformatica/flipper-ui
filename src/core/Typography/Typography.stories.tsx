import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Typography from '.'

export default {
    title: 'Typography',
    component: Typography
} as ComponentMeta<typeof Typography>

const Template: ComponentStory<typeof Typography> = args => (
    <Typography {...args}> {args.children} </Typography>
)

export const Default = Template.bind({})
Default.args = {
    children: 'This is a Typography text'
}

export const ErrorText = Template.bind({})
ErrorText.args = {
    color: 'error',
    children: 'Text with error color'
}

export const Primary = Template.bind({})
Primary.args = {
    color: 'primary',
    children: 'Text with primary color'
}

export const Secondary = Template.bind({})
Secondary.args = {
    color: 'secondary',
    children: 'Text with secondary color'
}

export const TextPrimary = Template.bind({})
TextPrimary.args = {
    color: 'textPrimary',
    children: 'Text with textPrimary color'
}

export const TextSecondary = Template.bind({})
TextSecondary.args = {
    color: 'textSecondary',
    children: 'Text with textSecondary color'
}

export const Button = Template.bind({})
Button.args = {
    variant: 'button',
    children: 'Button'
}

export const Caption = Template.bind({})
Caption.args = {
    variant: 'caption',
    children: 'Caption'
}

export const Body1 = Template.bind({})
Body1.args = {
    variant: 'body1',
    children: 'Body1'
}

export const Body2 = Template.bind({})
Body2.args = {
    variant: 'body2',
    children: 'Body2'
}

export const Subtitle1 = Template.bind({})
Subtitle1.args = {
    variant: 'subtitle1',
    children: 'Subtitle1'
}

export const Subtitle2 = Template.bind({})
Subtitle2.args = {
    variant: 'subtitle2',
    children: 'Subtitle2'
}

export const h1 = Template.bind({})
h1.args = {
    variant: 'h1',
    children: 'h1'
}

export const h2 = Template.bind({})
h2.args = {
    variant: 'h2',
    children: 'h2'
}

export const h3 = Template.bind({})
h3.args = {
    variant: 'h3',
    children: 'h3'
}

export const h4 = Template.bind({})
h4.args = {
    variant: 'h4',
    children: 'h4'
}

export const h5 = Template.bind({})
h5.args = {
    variant: 'h5',
    children: 'h5'
}

export const h6 = Template.bind({})
h6.args = {
    variant: 'h6',
    children: 'h6'
}
