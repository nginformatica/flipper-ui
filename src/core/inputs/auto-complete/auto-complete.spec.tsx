/* eslint-disable max-lines */
import * as React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import AutoComplete, { ISelected, TSelected } from '.'
import ListItem from '@/core/data-display/list-item'
import TextField from '../text-field'
import Typography from '@/core/data-display/typography'
import { act } from 'react-dom/test-utils'

interface MockProps {
    initialValue?: TSelected
    onFocus?: () => void
    focusDelay?: number
    onChange?: (value: TSelected) => void
    autoFocus?: boolean
    selectTextOnFocus?: boolean
    suggestions: ISelected[]
    openOnFocus?: boolean
    caseSensitive?: boolean
}

const MockAtuoComplete = (props: MockProps) => {
    const { suggestions, ...rest } = props
    const [value, setValue] = React.useState<TSelected>(
        props.initialValue ? props.initialValue : ''
    )

    const handleChange = (newValue: TSelected) => {
        props.onChange?.(newValue)
        setValue(newValue)
    }

    return (
        <AutoComplete
            value={value}
            suggestions={suggestions}
            renderInput={props => {
                return (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )
            }}
            renderSuggestion={(item, props, selected) => {
                const label = typeof item === 'string' ? item : item.label
                const hasSubheader = typeof item === 'object' && item.subheader

                return hasSubheader ? (
                    <ListItem onClick={props.onClick}>
                        <Typography variant='caption'>{item.label}</Typography>
                    </ListItem>
                ) : (
                    <ListItem onClick={props.onClick} selected={selected}>
                        {label}
                    </ListItem>
                )
            }}
            {...rest}
            onChange={handleChange}
        />
    )
}

describe('AutoComplete', () => {
    const SUGGESTIONS = [
        { value: 'rock', label: 'Rock' },
        { value: 'pop', label: 'Pop' },
        { value: 'classic', label: 'Classic' }
    ]

    const SUGGESTIONS_WITH_SUBHEADER = [
        { label: 'Genres', subheader: true },
        ...SUGGESTIONS
    ]
    it('should render suggestions', () => {
        render(
            <MockAtuoComplete
                openOnFocus
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const item1 = screen.getByText('Rock')
        const item2 = screen.getByText('Pop')
        const item3 = screen.getByText('Classic')

        expect(item1).toBeTruthy()
        expect(item2).toBeTruthy()
        expect(item3).toBeTruthy()
    })

    it('should render suggestions with case sensitive', async () => {
        render(
            <MockAtuoComplete
                caseSensitive
                openOnFocus
                suggestions={SUGGESTIONS}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        await act(async () => {
            await userEvent.type(input, 'R')
        })

        const item = screen.getByText('Rock')

        expect(item).toBeTruthy()
    })

    it('should render with no suggestions', () => {
        render(<MockAtuoComplete suggestions={[]} />)
        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const item1 = screen.queryByText('Rock')
        const item2 = screen.queryByText('Pop')
        const item3 = screen.queryByText('Classic')

        expect(item1).toBeFalsy()
        expect(item2).toBeFalsy()
        expect(item3).toBeFalsy()
    })

    it('should open with auto focus', async () => {
        render(
            <MockAtuoComplete
                autoFocus
                openOnFocus
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
                initialValue={{ label: 'Rock', value: 'rock' }}
            />
        )

        const subheader = await screen.findByText('Genres')
        const input = await screen.findByText('Rock')

        expect(subheader).toBeTruthy()
        expect(input).toBeTruthy()
    })

    it('should select suggestions', async () => {
        const onChangeSpy = jest.fn()
        render(
            <MockAtuoComplete
                openOnFocus
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
                initialValue={{ label: 'Rock', value: 'rock' }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const item1 = await screen.findByText('Rock')
        const item2 = screen.queryByText('Pop')
        const item3 = screen.queryByText('Classic')

        expect(item1).toBeTruthy()
        expect(item2).toBeFalsy()
        expect(item3).toBeFalsy()
    })

    it('should call onChange with first suggestion for default', () => {
        const onChangeSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)
        fireEvent.keyDown(input, { key: 'Enter' })

        expect(onChangeSpy).toHaveBeenCalledWith({
            value: 'rock',
            label: 'Rock'
        })
    })

    it('should navigate suggestions with arrow key down', () => {
        const onChangeSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)
        fireEvent.keyDown(input, { key: 'ArrowDown' })
        fireEvent.keyDown(input, { key: 'ArrowDown' })
        fireEvent.keyDown(input, { key: 'Enter' })

        expect(onChangeSpy).toHaveBeenCalledWith({
            value: 'classic',
            label: 'Classic'
        })
    })

    it('should navigate suggestions with arrow key down and up', () => {
        const onChangeSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)
        fireEvent.keyDown(input, { key: 'ArrowDown' })
        fireEvent.keyDown(input, { key: 'ArrowUp' })
        fireEvent.keyDown(input, { key: 'ArrowDown' })
        fireEvent.keyDown(input, { key: 'Enter' })

        expect(onChangeSpy).toHaveBeenCalledWith({ value: 'pop', label: 'Pop' })
    })

    it('should close suggestions on escape', async () => {
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={jest.fn()}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const beforeItem1 = screen.getByText('Rock')
        const beforeItem2 = screen.getByText('Pop')
        const beforeItem3 = screen.getByText('Classic')

        fireEvent.keyDown(input, { key: 'Escape' })

        await waitFor(() => {
            const afterItem1 = screen.queryByText('Rock')
            const afterItem2 = screen.queryByText('Pop')
            const afterItem3 = screen.queryByText('Classic')

            expect(beforeItem1).toBeDefined()
            expect(beforeItem2).toBeDefined()
            expect(beforeItem3).toBeDefined()

            expect(afterItem1).toBeNull()
            expect(afterItem2).toBeNull()
            expect(afterItem3).toBeNull()
        })
    })

    it('when typing should call onChange with correct values', async () => {
        const onChangeSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        userEvent.type(input, 'a')

        await waitFor(() => {
            expect(onChangeSpy).toHaveBeenCalledWith('a')
        })
    })

    it('should call onBlur on leave focus', () => {
        jest.useFakeTimers()
        const onBlurSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onBlur={onBlurSpy}
                onChange={jest.fn()}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.blur(input)
        jest.runAllTimers()

        const afterItem1 = screen.queryByText('Rock')

        expect(onBlurSpy).toHaveBeenCalled()
        expect(afterItem1).toBeNull()

        jest.useRealTimers()
    })

    it('should render suggestions with fade', () => {
        render(
            <AutoComplete
                openOnFocus
                fade
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={jest.fn()}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const item1 = screen.getByText('Rock')
        const item2 = screen.getByText('Pop')
        const item3 = screen.getByText('Classic')

        expect(item1).toBeTruthy()
        expect(item2).toBeTruthy()
        expect(item3).toBeTruthy()
    })

    it('should get suggestions with default width', () => {
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => {
                    props.inputProps.ref = jest.fn().mockReturnValue({
                        current: null
                    })

                    return (
                        <TextField
                            fullWidth
                            placeholder='input-placeholder'
                            {...props}
                        />
                    )
                }}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={jest.fn()}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const suggestionContainer = screen.getByRole('mui-paper-container')

        expect(suggestionContainer).toHaveProperty('style.width', '256px')
    })

    it('should call getBoundingClientRect when getPaperPosition is Above', () => {
        render(
            <AutoComplete
                openOnFocus
                maxHeight={999999}
                suggestions={SUGGESTIONS}
                renderInput={props => {
                    return (
                        <TextField
                            fullWidth
                            placeholder='input-placeholder'
                            {...props}
                        />
                    )
                }}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={jest.fn()}
            />
        )

        const input = screen.getByPlaceholderText('input-placeholder')

        fireEvent.focus(input)

        const suggestionContainer = screen.getByRole('mui-paper-container')

        expect(suggestionContainer).toHaveProperty('style.bottom', '1px')
    })

    it('should call onFocus', () => {
        const onFocusSpy = jest.fn()
        render(
            <MockAtuoComplete
                onFocus={onFocusSpy}
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
            />
        )

        const input = screen.getByPlaceholderText(
            'input-placeholder'
        ) as HTMLInputElement

        fireEvent.focus(input)

        expect(onFocusSpy).toBeCalled()
    })

    it('should selects a suggestion when clicking on it', () => {
        const onChangeSpy = jest.fn()
        render(
            <AutoComplete
                openOnFocus
                suggestions={SUGGESTIONS}
                renderInput={props => (
                    <TextField
                        fullWidth
                        placeholder='input-placeholder'
                        {...props}
                    />
                )}
                renderSuggestion={(item, props, selected) => {
                    const label = typeof item === 'string' ? item : item.label
                    const hasSubheader =
                        typeof item === 'object' && item.subheader

                    return hasSubheader ? (
                        <ListItem>
                            <Typography variant='caption'>
                                {item.label}
                            </Typography>
                        </ListItem>
                    ) : (
                        <ListItem {...props} selected={selected}>
                            {label}
                        </ListItem>
                    )
                }}
                onChange={onChangeSpy}
            />
        )

        const input = screen.getByPlaceholderText(
            'input-placeholder'
        ) as HTMLInputElement

        fireEvent.focus(input)

        const item = screen.getByText('Rock')

        fireEvent.click(item)

        expect(onChangeSpy).toHaveBeenCalledWith({
            value: 'rock',
            label: 'Rock'
        })
    })

    it('should update input value', async () => {
        render(<MockAtuoComplete suggestions={SUGGESTIONS_WITH_SUBHEADER} />)

        const input = screen.getByPlaceholderText(
            'input-placeholder'
        ) as HTMLInputElement

        await act(async () => {
            await userEvent.type(input, 'R')
        })

        const item = await waitFor(() => screen.findByText('Rock'))
        expect(input.value).toBe('R')
        expect(item).toBeTruthy()
    })

    it('should cancel if input value is a subheader', async () => {
        render(
            <MockAtuoComplete
                openOnFocus
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
            />
        )

        const input = screen.getByPlaceholderText(
            'input-placeholder'
        ) as HTMLInputElement

        fireEvent.focus(input)

        const item = await waitFor(() => screen.findByText('Genres'))

        await act(async () => {
            await userEvent.click(item)
        })

        expect(input.value).toBe('')
    })

    it('should select when selectTextOnFocus', async () => {
        render(
            <MockAtuoComplete
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
                selectTextOnFocus
            />
        )

        const input = screen.getByPlaceholderText(
            'input-placeholder'
        ) as HTMLInputElement

        await act(async () => await userEvent.type(input, 'Rock'))

        fireEvent.focus(input)

        expect(input.value).toBe('Rock')
    })

    it('should open suggestions with focusDelay', async () => {
        render(
            <MockAtuoComplete
                openOnFocus
                suggestions={SUGGESTIONS_WITH_SUBHEADER}
                focusDelay={100}
            />
        )

        const item = await waitFor(() => screen.findByText('Rock'))

        expect(item).toBeTruthy()
    })
})
