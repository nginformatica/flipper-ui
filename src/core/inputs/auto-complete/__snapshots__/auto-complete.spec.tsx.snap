// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete should match snapshot 1`] = `
<div>
  <div
    style="position: relative;"
  >
    <div
      class="sc-eqUzNf efRYEB"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        title=""
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          data-testid="text-field"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input makeStyles-outlinedInput-289"
            placeholder="input-placeholder"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-294 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-295"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
